src/main/kotlin/org/rust/ide/refactoring/inlineFunction/RsInlineFunctionProcessor.kt

```diff
@@ -158,296 +225,504 @@ class RsInlineFunctionProcessor(
+    (patterns zip expressions).flatMap { (parameter, argument) ->
+        matchPatternAndExpression(parameter, argument)
+    }
 
-                // Covering a case in which ; isn't included in the expression, and parent is too wide.
-                // e.g. if RsExpr is surrounded by RsBlock going to the RsBlock won't help.
-                val actualRetExpr: PsiElement = if (retExpr.parent.text == retExpr.text + ";") {
-                    retExpr.parent
-                } else {
-                    retExpr
-                }
-                if (caller.parent !is RsBlock && caller.parent.text != caller.text + ";") {
-                    caller.replace(retExprInside)
-                    retExpr.delete()
-                } else {
-                    actualRetExpr.replace(retExprInside)
-                    if (caller.parent.text == caller.text + ";") {
-                        caller.parent.delete()
-                    } else {
-                        caller.delete()
-                    }
-                }
-            }
-        } else {
-            caller.delete()
+private fun matchPatternAndExpression(pattern: RsPat, expression: RsExpr): List<Pair<RsPat, RsExpr>> {
+    if (pattern is RsPatTup && expression is RsTupleExpr) {
+        val patList = pattern.patList
+        val exprList = expression.exprList
+        if (patList.size == exprList.size) {
+            return matchPatternsAndExpressions(patList, exprList)
         }
     }
+    // TODO: RsPatRef etc
+    return listOf(pattern to expression)
+}
 
-    private fun replaceSelfParamWithExpr(selfParam: RsSelfParameter?, caller: PsiElement, funcScope: LocalSearchScope) {
-        if (selfParam != null) {
-            val selfExprText = buildString {
-                if (selfParam.mut != null) append("mut")
-                append(" ")
-                val callerText = when {
-                    caller is RsDotExpr -> caller.expr.text
-                    caller is RsCallExpr && caller.expr is RsPathExpr ->
-                        (caller as RsCallExprImpl).valueArgumentList.exprList[0].text
-                    else -> caller.text
-                }
-                append(callerText)
-            }
+private fun RsElement.containingLetDeclaration(): RsLetDecl? =
+    when (val parent = parent) {
+        is RsLetDecl -> parent
+        is RsPat -> parent.containingLetDeclaration()
+        else -> null
+    }
 
-            val selfExpr = factory.tryCreateExpression(selfExprText)!!
-            ReferencesSearch.search(selfParam.navigationElement, funcScope).findAll().forEach {
-                it.element.ancestorOrSelf<RsExpr>()!!.replace(selfExpr)
+private fun RsExpr.isSimpleLiteral(usagesCount: Int): Boolean {
+    return when (this) {
+        is RsUnaryExpr -> expr?.isSimpleLiteral(usagesCount) == true
+        is RsLitExpr -> {
+            if (usagesCount == 1) return true
```

Comment:
Consider replacing `usagesCount: Int` with `singleUse: Boolean` as it is only checked for `usagesCount == 1`.
By the way, is `usagesCount == 0` case possible? At least, I see `usages.isEmpty() && !argument.hasSideEffects` check in `trySubstituteParameterDirectly` but I did not find such tests where `usagesCount == 0` and it looks like this case is not covered.